MobileWeather/MainViewController.m

```diff
@@ -76,76 +57,63 @@ - (void)dealloc {
-    WeatherLocation *location = [[notification userInfo] objectForKey:@"location"];
+    WeatherLocation *location = notification.userInfo[@"location"];
     [self setLocationLabelFromWeatherLocation:location];
 }
 
 - (void)handleUnitUpdate:(NSNotification *)notification {
     WeatherDataManager *data = [WeatherDataManager sharedManager];
-    [self setWeatherConditions:[data weatherConditions] withLanguage:[data language]];
+    [self setWeatherConditions:data.weatherConditions withLanguage:data.language];
 }
 
 - (void)setLocationLabelFromWeatherLocation:(WeatherLocation *)location {
     NSString *locationText;
     
     if (location) {
-        locationText = [NSString stringWithFormat:@"%@, %@", [location country], [location city]];
-    }
-    else {
+        locationText = [NSString stringWithFormat:@"%@, %@, %@", location.city, location.state, location.country];
+    } else {
         locationText = @"...";
     }
-    
-    [[self locationLabel] setText:locationText];
+
+    self.locationLabel.text = locationText;
 }
 
 - (void)handleWeatherUpdate:(NSNotification *)notification {
-    NSDictionary *userInfo = [notification userInfo];
-    WeatherConditions *conditions = [userInfo objectForKey:@"weatherConditions"];
-    WeatherLanguage *language = [userInfo objectForKey:@"language"];
-    [self setWeatherConditions:conditions withLanguage:language];
+    NSDictionary *userInfo = notification.userInfo;
+    WeatherConditions *conditions = userInfo[@"weatherConditions"];
+    WeatherLanguage *language = userInfo[@"language"];
+
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self setWeatherConditions:conditions withLanguage:language];
+    });
 }
 
-- (void)setWeatherConditions:(WeatherConditions *)conditions withLanguage:(WeatherLanguage *)langugae {
-    NSDate *date = [conditions date];
+- (void)setWeatherConditions:(WeatherConditions *)conditions withLanguage:(WeatherLanguage *)language {
+    NSDate *date = conditions.date;
 
     NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
-    [formatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0]];
-    [formatter setLocale:[NSLocale localeWithLocaleIdentifier:[langugae value]]];
-    [formatter setDateStyle:NSDateFormatterFullStyle];
-    [formatter setTimeStyle:NSDateFormatterMediumStyle];
+    formatter.timeZone = [NSTimeZone timeZoneForSecondsFromGMT:0];
+    formatter.locale = [NSLocale localeWithLocaleIdentifier:language.value];
+    formatter.dateStyle = NSDateFormatterFullStyle;
+    formatter.timeStyle = NSDateFormatterMediumStyle;
     
     NSString *status = [formatter stringFromDate:date];
+
+    self.statusLabel.text = status;
+    self.currentConditionsLabel.text = conditions.conditionTitle;
+    self.conditionIcon.image = [[ImageProcessor sharedProcessor] imageFromConditionImage:conditions.conditionIcon];
+    self.precipitationChanceLabel.text = [conditions.precipitation stringValueForUnit:UnitPercentageDefault shortened:YES];
```

Comment:
Getting main thread warnings when
`self.statusLabel.text = status;` 
and 
`self.conditionIcon.image = [[ImageProcessor sharedProcessor] imageFromConditionImage:conditions.conditionIcon];`
are set.